pkg/controller/controller_utils.go

```diff
@@ -25,83 +26,97 @@ import (
+	// specifically targeted at the case where some problem prevents an update
+	// of expectations, without it the RC could stay asleep forever. This should
+	// be set based on the expected latency of watch events.
+	//
+	// Currently an rc can service (create *and* observe the watch events for said
+	// creation) about 10-20 pods a second, so it takes about 1 min to service
+	// 500 pods. Just creation is limited to 20qps, and watching happens with ~10-30s
+	// latency/pod at the scale of 3000 pods over 100 nodes.
+	ExpectationsTimeout = 3 * time.Minute
+)
+
+var (
+	KeyFunc = framework.DeletionHandlingMetaNamespaceKeyFunc
 )
 
-// Expectations are a way for replication controllers to tell the rc manager what they expect. eg:
-//	RCExpectations: {
-//		rc1: expects  2 adds in 2 minutes
-//		rc2: expects  2 dels in 2 minutes
-//		rc3: expects -1 adds in 2 minutes => rc3's expectations have already been met
+// Expectations are a way for controllers to tell the controller manager what they expect. eg:
+//	ControllerExpectations: {
+//		controller1: expects  2 adds in 2 minutes
+//		controller2: expects  2 dels in 2 minutes
+//		controller3: expects -1 adds in 2 minutes => controller3's expectations have already been met
 //	}
 //
 // Implementation:
 //	PodExpectation = pair of atomic counters to track pod creation/deletion
-//	RCExpectationsStore = TTLStore + a PodExpectation per rc
+//	ControllerExpectationsStore = TTLStore + a PodExpectation per controller
 //
 // * Once set expectations can only be lowered
-// * An RC isn't synced till its expectations are either fulfilled, or expire
-// * Rcs that don't set expectations will get woken up for every matching pod
+// * A controller isn't synced till its expectations are either fulfilled, or expire
+// * Controllers that don't set expectations will get woken up for every matching pod
 
-// expKeyFunc to parse out the key from a PodExpectation
-var expKeyFunc = func(obj interface{}) (string, error) {
+// ExpKeyFunc to parse out the key from a PodExpectation
+var ExpKeyFunc = func(obj interface{}) (string, error) {
 	if e, ok := obj.(*PodExpectations); ok {
 		return e.key, nil
 	}
 	return "", fmt.Errorf("Could not find key for obj %#v", obj)
 }
 
-// RCExpectationsManager is an interface that allows users to set and wait on expectations.
+// ControllerExpectationsInterface is an interface that allows users to set and wait on expectations.
 // Only abstracted out for testing.
-type RCExpectationsManager interface {
-	GetExpectations(rc *api.ReplicationController) (*PodExpectations, bool, error)
-	SatisfiedExpectations(rc *api.ReplicationController) bool
-	DeleteExpectations(rcKey string)
-	ExpectCreations(rc *api.ReplicationController, adds int) error
-	ExpectDeletions(rc *api.ReplicationController, dels int) error
-	CreationObserved(rc *api.ReplicationController)
-	DeletionObserved(rc *api.ReplicationController)
+type ControllerExpectationsInterface interface {
```

Comment:
It is not safe to use a single ControllerExpectationsInterface with different types of controllers because `framework.DeletionHandlingMetaNamespaceKey` may conflict across types. Might be worth adding a comment now.